sdk/core/azure-core/inc/azure/core/http/websockets/websockets.hpp

```diff
@@ -0,0 +1,368 @@
+       */
+      void Close(Azure::Core::Context const& context = Azure::Core::Context{});
+
+      /** @brief Closes a WebSocket connection to the remote server with additional context.
+       *
+       * @param closeStatus 16 bit WebSocket error code.
+       * @param closeReason String describing the reason for closing the socket.
+       * @param context Context for the operation.
+       */
+      void Close(
+          uint16_t closeStatus,
+          std::string const& closeReason = {},
+          Azure::Core::Context const& context = Azure::Core::Context{});
+
+      /** @brief Sends a String frame to the remote server.
+       *
+       * @param textFrame UTF-8 encoded text to send.
+       * @param isFinalFrame if True, this is the final frame in a multi-frame message.
+       * @param context Context for the operation.
+       */
+      void SendFrame(
+          std::string const& textFrame,
+          bool isFinalFrame = false,
+          Azure::Core::Context const& context = Azure::Core::Context{});
+
+      /** @brief Sends a Binary frame to the remote server.
+       *
+       * @param binaryFrame Binary data to send.
+       * @param isFinalFrame if True, this is the final frame in a multi-frame message.
+       * @param context Context for the operation.
+       */
+      void SendFrame(
+          std::vector<uint8_t> const& binaryFrame,
+          bool isFinalFrame = false,
+          Azure::Core::Context const& context = Azure::Core::Context{});
+
+      /** @brief Receive a frame from the remote server.
+       *
+       * @param context Context for the operation.
+       *
+       * @returns The received WebSocket frame.
+       *
+       */
+      std::shared_ptr<WebSocketFrame> ReceiveFrame(
+          Azure::Core::Context const& context = Azure::Core::Context{});
+
+      /** @brief AddHeader - Adds a header to the initial handshake.
+       *
+       * @note This API is ignored after the WebSocket is opened.
+       *
+       * @param headerName Name of header to add to the initial handshake request.
+       * @param headerValue Value of header to add.
+       */
+      void AddHeader(std::string const& headerName, std::string const& headerValue);
+
+      /** @brief Determine if the WebSocket is open.
+       *
+       * @returns true if the WebSocket is open, false otherwise.
+       */
+      bool IsOpen() const;
+
+      /** @brief Returns "true" if the configured websocket transport
+       * supports websockets in the transport, or if the websocket implementation
+       * is providing websocket protocol support.
+       *
+       * @returns true if the websocket transport supports websockets natively.
+       */
+      bool HasNativeWebSocketSupport() const;
+
+      /** @brief Returns the protocol chosen by the remote server during the initial handshake
+       *
+       * @returns The protocol negotiated between client and server.
+       */
+      std::string const& GetChosenProtocol() const;
```

Comment:
Nit, GetServerProtocol()

Reply:
Is it an error to call this before the handshake?  If so add to the notes on expected usage pattern

Reply:
It's not really the server protocol. The client offers up a list of protocols and the server choses one of them. This property expresses the protocol which was chosen by the server.

But I think that `GetNegotiatedProtocol()` may make more sense.